spec/validations/source_spec.rb

```diff
@@ -1,139 +1,47 @@
 # frozen_string_literal: true
 
 require 'spec_helper'
+require 'json'
 
 describe ZendeskAppsSupport::Validations::Source do
-  let(:files) { [double('AppFile', relative_path: 'abc.js')] }
-  let(:manifest_json) { { 'requirementsOnly' => false } }
-  let(:package) do
-    double('Package', js_files: files,
-                      lib_files: [],
-                      template_files: [],
-                      app_css: '',
-                      manifest: ZendeskAppsSupport::Manifest.new(JSON.dump(manifest_json)),
-                      locations: { 'zendesk' => { 'ticket_sidebar' => '_legacy' } },
-                      iframe_only?: false)
-  end
-
-  context 'when requirements only' do
-    let(:manifest_json) { { 'requirementsOnly' => true } }
+  let(:validation_error) { ZendeskAppsSupport::Validations::ValidationError }
+  let(:package) { double('Package', js_files: [], template_files: [], app_css: '') }
 
-    it 'should have an error when app.js is present' do
-      errors = ZendeskAppsSupport::Validations::Source.call(package)
-
-      expect(errors.first.key).to eql :no_code_for_ifo_notemplate
-    end
+  after { subject.call(package) }
 
-    it 'should not have an error when app.js is not present' do
-      allow(package).to receive(:js_files) { [] }
-      errors = ZendeskAppsSupport::Validations::Source.call(package)
-
-      expect(errors).to be_empty
-    end
-  end
-
-  context 'when iframe only' do
+  context 'when validating requirements_only app' do
     before do
-      allow(package.manifest).to receive(:iframe_only?) { true }
+      requirements_only_manifest =
+        ZendeskAppsSupport::Manifest.new({ requirementsOnly: true, marketingOnly: false }.to_json)
+      expect(package).to receive(:manifest).and_return(requirements_only_manifest)
     end
 
-    context 'when the package includes app.js' do
-      before do
-        allow(package).to receive(:js_files) { files }
-      end
+    it 'returns error if package contains source files' do
+      package.js_files << 'random_js_file.js'
 
-      it 'should have an error' do
-        errors = ZendeskAppsSupport::Validations::Source.call(package)
-        expect(errors.first.key).to eql :no_code_for_ifo_notemplate
-      end
+      expect(validation_error).to receive(:new).with(:no_code_for_ifo_notemplate)
     end
 
-    context 'when the package includes a library' do
-      before do
-        allow(package).to receive(:js_files) { [double('AppFile', relative_path: 'lib/slapp.js')] }
-      end
-
-      it 'should have an error' do
-        errors = ZendeskAppsSupport::Validations::Source.call(package)
-        expect(errors.first.key).to eql :no_code_for_ifo_notemplate
-      end
+    it 'returns nothing if package has no source files' do
+      expect(validation_error).not_to receive(:new).with(:no_code_for_ifo_notemplate)
```

Comment:
Would dropping the `.with(:no_code_for_ifo_notemplate)` condition here make this a better test? You'd never expect an error (of any kind) in this case right?

Also I found this a little odd to read with the expectation the only thing inside the `it` block. It's a bit redundant, but for readability maybe it's worth explicitly assigning `package.js_files = []` (or `package.files = []`)?

Reply:
Yeap let me address that.